File: php/oneapi/mms/MMSSend.php
<?php

require_once('Mail.php');
require_once('Mail/Mime.php');

require_once($ONEAPICOMMON.'foundation/FormParameter.php');
require_once($ONEAPICOMMON.'foundation/FormParameters.php');
require_once($ONEAPICOMMON.'foundation/JSONRequest.php');
require_once($ONEAPICOMMON.'response/RequestError.php');
require_once($ONEAPICOMMON.'response/ServiceException.php');
require_once($ONEAPICOMMON.'response/PolicyException.php');
require_once($ONEAPICOMMON.'response/ResourceReference.php');
require_once($ONEAPICOMMON.'response/mms/DeliveryInfo.php');
require_once($ONEAPICOMMON.'response/mms/DeliveryInfoList.php');
require_once($ONEAPICOMMON.'response/mms/CallbackReference.php');
require_once($ONEAPICOMMON.'response/mms/DeliveryReceiptSubscription.php');
require_once($ONEAPICOMMON.'response/mms/MMSSendResponse.php');
require_once($ONEAPICOMMON.'response/mms/MMSSendDeliveryStatusResponse.php');
require_once($ONEAPICOMMON.'response/mms/MMSDeliveryReceiptSubscriptionResponse.php');
	
class MMSSend {
	protected $endpoint;
	protected $username;
	protected $password;
	
	public function __construct ($endpoint = null, $username = null, $password = null) {
		$this->endpoint = $endpoint;
		$this->username = $username; 
		$this->password = $password;		
	}
	
	public function sendMMS($senderAddress, $addressList, $message, $attachments, $senderName, $clientCorrelator, $notifyURL, $callbackData) {
		$sendMMSResponse=null;
		if ($senderAddress && $addressList && ($message || $attachments)) {
			if (!is_null($this->endpoint) && property_exists($this->endpoint, 'SendSMS')) {
				$url=$this->endpoint->getSendMMSEndpoint();
			} else {
				$url='http://localhost:8080/oneapiserver/SendMMSService/1/messaging/outbound/{senderAddress}/requests';
			}
			
			$url=str_replace('{senderAddress}', urlencode($senderAddress), $url);
			
			$formParameters=new FormParameters();
			$formParameters->put("senderAddress", $senderAddress);
			for ($i=0; $i<count($addressList); $i++) $formParameters->put("address", $addressList[$i]); 			
			$formParameters->put("message", $message);
			$formParameters->put("clientCorrelator", $clientCorrelator);
			$formParameters->put("notifyURL", $notifyURL);
			$formParameters->put("senderName", $senderName);			
			$formParameters->put("callbackData", $callbackData);	
			
			$params=array();		
			$params['content_type'] = 'multipart/mixed';
			$requestBody = new Mail_mimePart('', $params);

			$params['content_type'] = 'application/x-www-form-urlencoded';
			//$params['encoding']     = '8bit';
			$params['disposition']    = 'form-data; name="root-fields"';
			$text = $requestBody->addSubPart($formParameters->encodeParameters(), $params);
			
			error_log('Processing attachments');
			if ($attachments && is_array($attachments) && count($attachments)>0) {
				for ($i=0; $i<count($attachments); $i++) {
					error_log('Processing file '.$i);
					$file=$attachments[$i];
					if ($file && is_array($file) && isset($file['error']) && $file['error']==UPLOAD_ERR_OK) {
						error_log('Have attachment name='.$file['name'].' type='.$file['type'].' size='.$file['size']);
						$params['content_type'] = $file['type'];
						$params['encoding']     = 'base64';
						$params['disposition']  = 'form-data; name="'.$file['name'].'"';
						$contents=file_get_contents($file['tmp_name']);
						$attach =& $requestBody->addSubPart($contents, $params);
					} 
				}
			}
			
			$content=$requestBody->encode();	
			$contentType=$content['headers']['Content-Type'];
			$data=$content['body'];
			
			error_log('Generated...\n'.print_r($content,true));																																																
			
			$request=new JSONRequest($url, $this->username, $this->password, 'POST', null, $contentType, $data);
			$response=$request->execute();
		
			if ($response->getResponseBody()) {
				$jsondata=json_decode($response->getResponseBody());
			}
			
			error_log("MMSSend::sendMMS response=".print_r($response,true));
			error_log("MMSSend::sendMMS jsondata=".print_r($jsondata,true));
			
			$responseInfo=$response->getResponseInfo();
			$location=$response->getLocation();
	
			$sendMMSResponse=new MMSSendResponse($responseInfo["http_code"], $responseInfo["content_type"], $location, $jsondata);
			
			error_log("MMSSend::sendMMS sendMMSResponse=".print_r($sendMMSResponse,true));
		}
		
		return $sendMMSResponse;
	}

	public function queryDeliveryStatus($senderAddress, $requestId) {
		$mmsSendDeliveryStatusResponse=null;
		if ($senderAddress && $requestId) {
			if (!is_null($this->endpoint) && property_exists($this->endpoint, 'QuerySMSDelivery')) {
				$url=$this->endpoint->getQueryMMSDeliveryEndpoint();
			} else {
				$url='http://localhost:8080/oneapiserver/QueryMMSService/1/messaging/outbound/{senderAddress}/requests/{requestId}/deliveryInfos';
			}
			
			$url=str_replace('{senderAddress}', urlencode($senderAddress), $url);
			$url=str_replace('{requestId}', urlencode($requestId), $url);
			
			$request=new JSONRequest($url, $this->username, $this->password, 'GET', null);
			$response=$request->execute();
		
			if ($response->getResponseBody()) {
				$jsondata=json_decode($response->getResponseBody());
			}
			
			error_log("MMSSend::queryDeliveryStatus response=".print_r($response,true));
			error_log("MMSSend::queryDeliveryStatus jsondata=".print_r($jsondata,true));
			
			$responseInfo=$response->getResponseInfo();
			$location=$response->getLocation();
	
			$mmsSendDeliveryStatusResponse=new MMSSendDeliveryStatusResponse($responseInfo["http_code"], $responseInfo["content_type"], $location, $jsondata);
			
			error_log("MMSSend::queryDeliveryStatus mmsSendDeliveryStatusResponse=".print_r($mmsSendDeliveryStatusResponse,true));
		}
		
		return $mmsSendDeliveryStatusResponse;
	}
	
	public function subscribeToDeliveryNotifications($senderAddress, $clientCorrelator, $notifyURL, $callbackData) {
		$mmsDeliveryReceiptSubscriptionResponse=null;
		if ($senderAddress && $notifyURL) {
			if (!is_null($this->endpoint) && property_exists($this->endpoint, 'MMSDeliverySubscriptions')) {
				$url=$this->endpoint->getMMSDeliverySubscriptionsEndpoint();
			} else {
				$url='http://localhost:8080/oneapiserver/MMSDeliveryService/1/messaging/outbound/{senderAddress}/subscriptions';
			}
			
			$url=str_replace('{senderAddress}', urlencode($senderAddress), $url);
			
			$formParameters=new FormParameters();
			$formParameters->put("clientCorrelator", $clientCorrelator);
			$formParameters->put("notifyURL", $notifyURL);
			$formParameters->put("callbackData", $callbackData);			
			
			$request=new JSONRequest($url, $this->username, $this->password, 'POST', $formParameters->getParameterSet());
			$response=$request->execute();
		
			if ($response->getResponseBody()) {
				$jsondata=json_decode($response->getResponseBody());
			}
			
			error_log("MMSSend::subscribeToDeliveryNotifications response=".print_r($response,true));
			error_log("MMSSend::subscribeToDeliveryNotifications jsondata=".print_r($jsondata,true));
			
			$responseInfo=$response->getResponseInfo();
			$location=$response->getLocation();
	
			$mmsDeliveryReceiptSubscriptionResponse=new MMSDeliveryReceiptSubscriptionResponse($responseInfo["http_code"], $responseInfo["content_type"], $location, $jsondata);
			
			error_log("MMSSend::subscribeToDeliveryNotifications mmsDeliveryReceiptSubscriptionResponse=".print_r($mmsDeliveryReceiptSubscriptionResponse,true));
		}
		
		return $mmsDeliveryReceiptSubscriptionResponse;
	}
	
	public function cancelDeliveryNotifications($subscriptionId) {
		$responseCode=null;
		if ($subscriptionId) {
			if (!is_null($this->endpoint) && property_exists($this->endpoint, 'CancelMMSDeliverySubscription')) {
				$url=$this->endpoint->getCancelMMSDeliverySubscriptionEndpoint();
			} else {
				$url='http://localhost:8080/oneapiserver/CancelMMSDeliveryService/1/messaging/outbound/subscriptions/{subscriptionId}';
			}
			
			$url=str_replace('{subscriptionId}', urlencode($subscriptionId), $url);
			
			$request=new JSONRequest($url, $this->username, $this->password, 'DELETE', null);
			$response=$request->execute();
			$responseInfo=$response->getResponseInfo();

			error_log("MMSSend::cancelDeliveryNotifications httpResponseCode=".$responseInfo["http_code"]);
			$responseCode=$responseInfo["http_code"];

		}		
		
		return $responseCode;
	}
	
}

?>

File: oneapi/src/org/gsm/oneapi/mms/MMSSend.java
package org.gsm.oneapi.mms;

import java.io.ByteArrayOutputStream;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.UUID;

import javax.mail.internet.InternetHeaders;
import javax.mail.internet.MimeBodyPart;
import javax.mail.internet.MimeMultipart;

import org.apache.commons.fileupload.FileItem;
import org.apache.log4j.Logger;
import org.gsm.oneapi.endpoints.ServiceEndpoints;
import org.gsm.oneapi.foundation.FormParameters;
import org.gsm.oneapi.foundation.JSONRequest;
import org.gsm.oneapi.responsebean.RequestError;
import org.gsm.oneapi.responsebean.mms.MMSDeliveryReceiptSubscriptionResponse;
import org.gsm.oneapi.responsebean.mms.MMSSendDeliveryStatusResponse;
import org.gsm.oneapi.responsebean.mms.MMSSendResponse;
import org.gsm.oneapi.responsebean.sms.SMSDeliveryReceiptSubscriptionResponse;
import org.gsm.oneapi.responsebean.sms.SMSSendDeliveryStatusResponse;
import org.gsm.oneapi.responsebean.sms.SMSSendResponse;
import org.gsm.oneapi.server.OneAPIServlet;

public class MMSSend {

	static Logger logger=Logger.getLogger(MMSSend.class);
	
	ServiceEndpoints endPoints=null;
	String authorisationHeader=null;
	
	public static boolean dumpRequestAndResponse=false;

	private static JSONRequest<MMSSendResponse> mmsSendResponseProcessor=new JSONRequest<MMSSendResponse>(new MMSSendResponse());
	private static JSONRequest<MMSSendDeliveryStatusResponse> mmsSendDeliveryStatusProcessor=new JSONRequest<MMSSendDeliveryStatusResponse>(new MMSSendDeliveryStatusResponse());
	private static JSONRequest<MMSDeliveryReceiptSubscriptionResponse> mmsDeliveryReceiptSubscriptionProcessor=new JSONRequest<MMSDeliveryReceiptSubscriptionResponse>(new MMSDeliveryReceiptSubscriptionResponse());

	/**
	 Creates a new instance of the Send MMS API main interface. Requires endPoints to define the URL targets of the various Send MMS network calls and authorisationHeader containing the username/password used for HTTP Basic authorisation with the OneAPI server.  
	                          
	@param  endPoints  contains a set of service/ call specific endpoints 	
	@param  authorisationHeader Base 64 encoded username/ password
	     
	@see org.gsm.oneapi.endpoints.ServiceEndpoints
	@see org.gsm.oneapi.foundation.JSONRequest#getAuthorisationHeader(String, String)
	 */	
	public MMSSend(ServiceEndpoints endPoints, String authorisationHeader) {
		this.endPoints=endPoints;
		this.authorisationHeader=authorisationHeader;
	}
	
	/**
	 Can be used to update the service endpoints  
	                          
	@param  endPoints  contains a set of service/ call specific endpoints 	

	@see org.gsm.oneapi.endpoints.ServiceEndpoints
	 */	
	public void setEndpoints(ServiceEndpoints endPoints) {
		this.endPoints=endPoints;		
	}
	
	/**
	 Can be used to update the service authorisation header  
	                          
	@param  authorisationHeader Base 64 encoded username/ password

	@see org.gsm.oneapi.foundation.JSONRequest#getAuthorisationHeader(String, String)
	 */	
	public void setAuthorisationHeader(String authorisationHeader) {
		this.authorisationHeader=authorisationHeader;
	}

	/**
	 Send an MMS to one or more mobile terminals  
	                          
	@param senderAddress (mandatory) is the address to whom a responding MMS may be sent. Do not URL encode this value prior to passing to this function
	@param address (mandatory) contains at least one address for each end user ID to send to; in this case their MSISDN including the ‘tel:’ protocol identifier and the country code preceded by ‘+’. i.e., tel:+16309700001.  OneAPI also supports the Anonymous Customer Reference (ACR) if provided by the operator.
	@param message (mandatory) contains the message text to send
	@param senderName (mandatory) is the name to appear on the user’s terminal as the sender of the message
	@param attachments (optional) contains a list of attachments to be sent along with this MMS request
	@param clientCorrelator (optional) uniquely identifies this create MMS request. If there is a communication failure during the request, using the same clientCorrelator when retrying the request allows the operator to avoid sending the same MMS twice.
	@param notifyURL (optional) is the URL to which you would like a notification of delivery sent
	@param callbackData (optional) will be passed back to the notifyURL location, so you can use it to identify the message the receipt relates to (or any other useful data, such as a function name)
	     
	@see MMSSendResponse
	*/
	public MMSSendResponse sendMMS(String senderAddress, String[] address, String message, ArrayList<FileItem> attachments, String senderName, String clientCorrelator, String notifyURL, String callbackData) {
		MMSSendResponse response=new MMSSendResponse();
		
		FormParameters formParameters=new FormParameters();
		formParameters.put("senderAddress", senderAddress);
		for (String addr:address) if (address!=null) formParameters.put("address", addr);		
		formParameters.put("message", message);				
		formParameters.put("clientCorrelator", clientCorrelator);
		formParameters.put("notifyURL", notifyURL);
		formParameters.put("senderName", senderName);
		formParameters.put("callbackData", callbackData);
				
		if (true /* validation goes here */) {

			String endpoint=endPoints.getSendMMSEndpoint();
			
	    	int responseCode=0;
	        String contentType = null;
			
			try {
		        MimeMultipart mp = new MimeMultipart("mixed");

		        InternetHeaders rootHdr = new InternetHeaders();

		        rootHdr.addHeader("Content-Type", "application/x-www-form-urlencoded");
		        rootHdr.addHeader("Content-Disposition", "form-data; name=\"root-fields\"");

		        byte[] rootData = JSONRequest.formEncodeLineSeparaterParams(formParameters).getBytes();
		        mp.addBodyPart(new MimeBodyPart(rootHdr, rootData));
		        
		        if (attachments!=null && attachments.size()>0) {
		        	for (int i=0; i<attachments.size(); i++) {
		        		FileItem attachment=attachments.get(i);
		        		if (attachment!=null) {
			        		logger.debug("Attaching file ["+i+"] name="+attachment.getName()+" type="+attachment.getContentType());
			        		InternetHeaders attachmentHdr = new InternetHeaders();
			        		if (attachment.getContentType()!=null) attachmentHdr.addHeader("Content-Type", attachment.getContentType());
			        		if (attachment.getName()!=null) attachmentHdr.addHeader("Content-Disposition", "form-data; name=\""+attachment.getName()+"\"");
			        		byte[] dataBytes=attachment.get();
			        		MimeBodyPart mpa=new MimeBodyPart(attachmentHdr, dataBytes);
			        		mp.addBodyPart(mpa);
		        		}
		        	}
		        }
			
		        ByteArrayOutputStream baos = new ByteArrayOutputStream();

		        mp.writeTo(baos);
		        
		        String content=baos.toString();
		        String boundary = content.substring(2, content.indexOf("\r\n"));

				
				if (dumpRequestAndResponse) JSONRequest.dumpRequestVariables(endpoint, authorisationHeader, formParameters); 
	
				HttpURLConnection con = JSONRequest.setupConnection(endpoint, authorisationHeader);
		        con.setRequestMethod("POST");
		        UUID separator=UUID.randomUUID();
		        
		        con.setRequestProperty("Content-Type", "multipart/form-data; boundary=\""+boundary+"\"");
		        con.setDoOutput(true);
		        
		        OutputStreamWriter out = new OutputStreamWriter(con.getOutputStream());
	            out.write(content);
	            out.close();
	            
//	            System.out.println("Sending:\n"+content);
//	            		        
	        	responseCode = con.getResponseCode();
	            contentType = con.getContentType();
		        
	            response=mmsSendResponseProcessor.getResponse(con, OneAPIServlet.CREATED);
			} catch (Exception e) {
				response.setHTTPResponseCode(responseCode);
				response.setContentType(contentType);
				
				response.setRequestError(new RequestError(RequestError.SERVICEEXCEPTION, "SVCJAVA", e.getMessage(), e.getClass().getName()));
	
				logger.error("Exception "+e.getMessage()+" "+e.getLocalizedMessage());
			}
		}
		return response;
	}
	
	/**
	 Query the delivery status for an MMS sent to one or more mobile terminals  
	                          
	@param senderAddress (mandatory) is the address from which MMS messages are being sent. Do not URL encode this value prior to passing to this function
	@param requestId (mandatory) contains the requestId returned from a previous call to the sendMMS function
	     
	@see MMSSendDeliveryStatusResponse
	*/
	public MMSSendDeliveryStatusResponse queryDeliveryStatus(String senderAddress, String requestId) {
		MMSSendDeliveryStatusResponse response=new MMSSendDeliveryStatusResponse();
		
		if (senderAddress!=null && requestId!=null) {
			
	    	int responseCode=0;
	        String contentType = null;

			try {
				logger.debug("endpoint="+endPoints.getQueryMMSDeliveryEndpoint());
				String endpoint=endPoints.getQueryMMSDeliveryEndpoint().replaceAll("\\{senderAddress\\}", URLEncoder.encode(senderAddress, "utf-8")).replaceAll("\\{requestId\\}", URLEncoder.encode(requestId, "utf-8"));
				
				if (dumpRequestAndResponse) JSONRequest.dumpRequestVariables(endpoint, authorisationHeader, null);
				
				HttpURLConnection con = JSONRequest.setupConnection(endpoint, authorisationHeader);
	
	        	responseCode=con.getResponseCode();
	            contentType = con.getContentType();
	            
	            response.setHTTPResponseCode(responseCode);
	            response.setContentType(contentType);
		        
	            response=mmsSendDeliveryStatusProcessor.getResponse(con, OneAPIServlet.OK);
			} catch (Exception e) {
				logger.error("Exception "+e.getMessage()+" "+e.getLocalizedMessage());
				e.printStackTrace();
				response.setHTTPResponseCode(responseCode);
				response.setContentType(contentType);
				
				response.setRequestError(new RequestError(RequestError.SERVICEEXCEPTION, "SVCJAVA", e.getMessage(), e.getClass().getName()));
			}
		}
		return response;
	}

	/**
	 Start subscribing to delivery status notifications for all your sent MMS  
	                          
	@param senderAddress (mandatory) is the address from which MMS messages are being sent. Do not URL encode this value prior to passing to this function
	@param notifyURL (mandatory) is the URL to which you would like a notification of delivery sent
	@param clientCorrelator (optional) uniquely identifies this subscription request. If there is a communication failure during the request, using the same clientCorrelator when retrying the request allows the operator to avoid setting up the same subscription twice.
	@param callbackData (optional) will be passed back to the notifyURL location, so you can use it to identify the message the delivery receipt relates to (or any other useful data, such as a function name)

	@see MMSDeliveryReceiptSubscriptionResponse
	*/
	public MMSDeliveryReceiptSubscriptionResponse subscribeToDeliveryNotifications(String senderAddress, String clientCorrelator, String notifyURL, String callbackData) {
		MMSDeliveryReceiptSubscriptionResponse response=new MMSDeliveryReceiptSubscriptionResponse();

		if (senderAddress!=null && notifyURL!=null) {
			FormParameters formParameters=new FormParameters();
			formParameters.put("clientCorrelator", clientCorrelator);
			formParameters.put("notifyURL", notifyURL);
			formParameters.put("callbackData", callbackData);
			
	    	int responseCode=0;
	        String contentType = null;
			
			try {
				String endpoint=endPoints.getMMSDeliverySubscriptionsEndpoint().replaceAll("\\{senderAddress\\}", URLEncoder.encode(senderAddress, "utf-8"));
				
				if (dumpRequestAndResponse) JSONRequest.dumpRequestVariables(endpoint, authorisationHeader, formParameters); 
	
				HttpURLConnection con = JSONRequest.setupConnection(endpoint, authorisationHeader);
		        con.setDoOutput(true);
		        OutputStreamWriter out = new OutputStreamWriter(con.getOutputStream());
		        
		        String requestBody=JSONRequest.formEncodeParams(formParameters);
		        out.write(requestBody);
		        out.close();
	
	        	responseCode=con.getResponseCode();
	            contentType = con.getContentType();
		        
	            response=mmsDeliveryReceiptSubscriptionProcessor.getResponse(con, OneAPIServlet.CREATED);
			} catch (Exception e) {
				response.setHTTPResponseCode(responseCode);
				response.setContentType(contentType);
				
				response.setRequestError(new RequestError(RequestError.SERVICEEXCEPTION, "SVCJAVA", e.getMessage(), e.getClass().getName()));
				logger.error("Exception "+e.getMessage()+" "+e.getLocalizedMessage());
			}
		}
		return response;
	}

	/**
	 Stop subscribing to delivery status notifications for all your sent MMS  
	                          
	@param subscriptionId (mandatory) contains the subscriptionId of a previously created MMS delivery receipt subscription

	*/
	public int cancelDeliveryNotifications(String subscriptionId)  {
		int responseCode=0;
		
		if (subscriptionId!=null) {			
			try {
				logger.debug("endpoint="+endPoints.getCancelMMSDeliverySubscriptionEndpoint());
				String endpoint=endPoints.getCancelMMSDeliverySubscriptionEndpoint().replaceAll("\\{subscriptionId\\}", URLEncoder.encode(subscriptionId, "utf-8"));
				
				if (dumpRequestAndResponse) JSONRequest.dumpRequestVariables(endpoint, authorisationHeader, null);
				
				HttpURLConnection con = JSONRequest.setupConnection(endpoint, authorisationHeader);
				
				con.setRequestMethod("DELETE");
	
	        	responseCode=con.getResponseCode();
			} catch (Exception e) {
				logger.error("Exception "+e.getMessage()+" "+e.getLocalizedMessage());
				e.printStackTrace();
			}
		}
		return responseCode;
	}



}
